File: Min_Heap_using_Binary_Tree.c
#include <stdio.h>
#include <stdlib.h>

typedef struct Node {
    int data;
    struct Node *left, *right, *parent;
} Node;

typedef struct { 
    Node* root;
} Heap;

typedef struct {
    Node** data;
    int front, rear, size;
} Queue;

Queue* createQueue(int cap) {
    Queue* queue = (Queue*)malloc(sizeof(Queue));
    queue->data = (Node**)malloc(sizeof(Node*) * cap);
    queue->front = 0;
    queue->rear = 0;
    queue->size = 0;
    return queue;
}

void enqueue(Queue* queue, Node* node) {
    queue->data[queue->rear++] = node;
    queue->size++;
}

Node* dequeue(Queue* queue) {
    queue->size--;
    return queue->data[queue->front++];
}

int isEmpty(Queue* queue) {return queue->size == 0;}

void freeQueue(Queue* queue) {
    free(queue->data);
    free(queue);
}

Node* createNode(int val) {
    Node* node = (Node*)malloc(sizeof(Node));
    node->data = val;
    node->left = node->right = node->parent = NULL;
    return node;
}

Node** getLevelOrder(Node* root, int* count) {
    *count = 0;
    if (!root) return NULL;

    Node** nodes = (Node**)malloc(sizeof(Node*) * 1000);
    Queue* queue = createQueue(1000);

    enqueue(queue, root);
    while (!isEmpty(queue)) {
        Node* current = dequeue(queue);
        nodes[(*count)++] = current;

        if (current->left) enqueue(queue, current->left);
        if (current->right) enqueue(queue, current->right);
    }
    freeQueue(queue);
    return nodes;
}

void percolateUp(Node* node) {
    while (node->parent && node->data < node->parent->data) {
        int temp = node->data;
        node->data = node->parent->data;
        node->parent->data = temp;
        node = node->parent;
    }
}

void heapify(Node* node) {
    if (!node) return;

    Node* smallest = node;
    if (node->left && node->left->data < smallest->data) smallest = node->left;
    if (node->right && node->right->data < smallest->data) smallest = node->right;

    if (smallest != node) {
        int temp = node->data;
        node->data = smallest->data;
        smallest->data = temp;
        heapify(smallest);
    }
}

Node* getLastNode(Node* root) {
    int count;
    Node** nodes = getLevelOrder(root, &count);
    Node* last = count > 0 ? nodes[count - 1] : NULL;
    free(nodes);
    return last;
}

void insertNode(Heap* heap, int val) {
    Node* newNode = createNode(val);
    if (!heap->root) {
        heap->root = newNode;
        return;
    }

    int count;
    Node** nodes = getLevelOrder(heap->root, &count);

    for (int i = 0; i < count; i++) {
        Node* node = nodes[i];
        if (!node->left) {
            node->left = newNode;
            newNode->parent = node;
            percolateUp(newNode);
            break;
        } else if (!node->right) {
            node->right = newNode;
            newNode->parent = node;
            percolateUp(newNode);
            break;
        }
    }
    free(nodes);
}

void insert(Heap* heap, int val) {insertNode(heap, val);}

void deleteMin(Heap* heap) {
    if (!heap->root) {
        printf("Heap is empty\n");
        return;
    }

    Node* last = getLastNode(heap->root);
    if (last == heap->root) {
        free(heap->root);
        heap->root = NULL;
        return;
    }

    heap->root->data = last->data;
    if (last->parent->left == last) last->parent->left = NULL;
    else last->parent->right = NULL;

    free(last);
    heapify(heap->root);
}

int findMin(Heap* heap) {
    if (!heap->root) {
        printf("Heap is empty\n");
        return -1;
    } return heap->root->data;
}

int findMax(Heap* heap) {
    int count;
    Node** nodes = getLevelOrder(heap->root, &count);
    if (count == 0) {
        printf("Heap is empty\n");
        return -1;
    }
    int maxVal = nodes[0]->data;
    for (int i = 1; i < count; i++) if (nodes[i]->data > maxVal) maxVal = nodes[i]->data;
    free(nodes);
    return maxVal;
}

void display(Heap* heap) {
    if (!heap->root) {
        printf("Heap is empty\n");
        return;
    }
    int count;
    Node** nodes = getLevelOrder(heap->root, &count);
    for (int i = 0; i < count; i++) printf("%d ", nodes[i]->data);
    printf("\n");
    free(nodes);
}

int main() {
    Heap heap;
    heap.root = NULL;

    printf("\nMin Heap Functions:\n1. Insert\n2. Delete Min\n3. Find Min\n4. Find Max\n5. Display\n");
    while (1) {
        int choice, elem;
        printf("\nEnter Heap Operation: ");
        scanf("%d", &choice);

        if (choice == 1) {
            printf("Enter element to insert: ");
            scanf("%d", &elem);
            insert(&heap, elem);
        }
        else if (choice == 2) deleteMin(&heap);
        else if (choice == 3) printf("Min: %d\n", findMin(&heap));
        else if (choice == 4) printf("Max: %d\n", findMax(&heap));
        else if (choice == 5) display(&heap);
        else printf("Invalid choice\n");
    }
    return 0;
}

File: Java/Question Practice Java/Min_Heap_using_Binary_Tree.java
import java.util.ArrayList;
import java.util.LinkedList;
import java.util.List;
import java.util.Queue;
import java.util.Scanner;

class Node {
    int data;
    Node left, right, parent;

    Node(int val) {
        data = val;
        left = right = parent = null;
    }
}

class Heap {
    private Node root;

    public Heap() {root = null;}

    private List<Node> getLevelOrder() {
        List<Node> nodes = new ArrayList<>();
        if (root == null) return nodes;

        Queue<Node> queue = new LinkedList<>();
        queue.add(root);
        while (!queue.isEmpty()) {
            Node curr = queue.poll();
            nodes.add(curr);
            if (curr.left != null) queue.add(curr.left);
            if (curr.right != null) queue.add(curr.right);
        }
        return nodes;
    }

    private void insertNode(int val) {
        Node newNode = new Node(val);
        if (root == null) {
            root = newNode;
            return;
        } else {
            List<Node> nodes = getLevelOrder();
            for (Node node : nodes) {
                if (node.left == null) {
                    node.left = newNode;
                    newNode.parent = node;
                    percolateUp(newNode);
                    return;
                } else if (node.right == null) {
                    node.right = newNode;
                    newNode.parent = node;
                    percolateUp(newNode);
                    return;
                }
            }
        }
    }

    private void percolateUp(Node node) {
        while (node.parent != null && node.data < node.parent.data) {
            int temp = node.data;
            node.data = node.parent.data;
            node.parent.data = temp;
            node = node.parent;
        }
    }

    private void heapify(Node node) {
        if (node == null) return;

        Node smallest = node;
        if (node.left != null && node.left.data < smallest.data) smallest = node.left;
        if (node.right != null && node.right.data < smallest.data) smallest = node.right;

        if (smallest != node) {
            int temp = node.data;
            node.data = smallest.data;
            smallest.data = temp;
            heapify(smallest);
        }
    }

    private Node getLastNode() {
        List<Node> nodes = getLevelOrder();
        return nodes.isEmpty() ? null : nodes.get(nodes.size() - 1);
    }

    public void insert(int val) {insertNode(val);}

    public void deleteMin() {
        if (root == null) System.out.println("Heap is empty");
        else {
            Node last = getLastNode();
            if (last == root) {
                root = null;
                return;
            }
            root.data = last.data;

            if (last.parent.left == last) last.parent.left = null;
            else last.parent.right = null;

            heapify(root);
        }
    }

    public int findMin() {
        if (root == null) {
            System.out.println("Heap is empty");
            return -1;
        } return root.data;
    }

    public int findMax() {
        List<Node> nodes = getLevelOrder();
        if (nodes.isEmpty()) {
            System.out.println("Heap is empty");
            return -1;
        }
        int maxVal = nodes.get(0).data;
        for (Node node : nodes) if (node.data > maxVal) maxVal = node.data;
        return maxVal;
    }

    public void display() {
        if (root == null) System.out.println("Heap is empty");
        else {
            List<Node> nodes = getLevelOrder();
            for (Node node : nodes) System.out.print(node.data + " ");
            System.out.println();
        }
    }
}

public class Min_Heap_using_Binary_Tree {
    public static void main(String[] args) {
        Scanner input = new Scanner(System.in);
        Heap heap = new Heap();
        System.out.println("Min Heap Functions:\n1. Insert\n2. Delete Min\n3. Find Min\n4. Find Max\n5. Display");

        while (true) {
            System.out.print("\nEnter Heap Operation: ");
            int choice = input.nextInt();

            if (choice == 1) {
                System.out.print("Enter element to insert: ");
                int elem = input.nextInt();
                heap.insert(elem);
            } else if (choice == 2) heap.deleteMin();
            else if (choice == 3) System.out.println("Min: " + heap.findMin());
            else if (choice == 4) System.out.println("Max: " + heap.findMax());
            else if (choice == 5) heap.display();
            else System.out.println("Invalid choice");
        }
    }
}